crates/sui-oracle/move/oracle/sources/meta_oracle.move

```diff
@@ -0,0 +1,260 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+module oracle::meta_oracle {
+    use std::option::{Self, Option};
+    use std::string::String;
+    use std::type_name;
+    use std::vector;
+
+    use oracle::data::{Self, Data};
+    use oracle::decimal_value::DecimalValue;
+    use oracle::simple_oracle::SimpleOracle;
+    use sui::bcs;
+    use sui::math;
+
+    #[test_only]
+    use oracle::decimal_value;
+
+    const EValidDataSizeLessThanThreshold: u64 = 0;
+    const EUnsupportedDataType: u64 = 1;
+
+    struct MetaOracle<T> {
+        oracle_datas: vector<Option<Data<T>>>,
+        threshole: u64,
+        time_window_ms: u64,
+        ticker: String,
+        max_timestamp: u64,
+    }
+
+    public fun new<T: copy + drop>(threshole: u64, time_window_ms: u64, ticker: String): MetaOracle<T> {
+        MetaOracle {
+            oracle_datas: vector::empty(),
+            threshole,
+            time_window_ms,
+            ticker,
+            max_timestamp: 0,
+        }
+    }
+
+    public fun add_simple_oracle<T: copy + drop + store>(meta_oracle: &mut MetaOracle<T>, oracle: &SimpleOracle) {
+        let oracle_data = oracle::simple_oracle::get_latest_data(oracle, meta_oracle.ticker);
+        if (option::is_some(&oracle_data)) {
+            meta_oracle.max_timestamp = data::timestamp(option::borrow(&oracle_data));
+        };
+        vector::push_back(&mut meta_oracle.oracle_datas, oracle_data);
+    }
+
+    struct TrustedData<T> has copy, drop {
+        value: T,
+        oracles: vector<address>,
+    }
+
+    fun combind<T: copy + drop>(meta_oracle: MetaOracle<T>, ): (vector<T>, vector<address>) {
+        let MetaOracle { oracle_datas, threshole, time_window_ms, ticker: _, max_timestamp } = meta_oracle;
+        let min_timestamp = max_timestamp - time_window_ms;
+        let values = vector<T>[];
+        let oracles = vector<address>[];
+        while (vector::length(&oracle_datas) > 0) {
+            let oracle_data = vector::remove(&mut oracle_datas, 0);
+            if (option::is_some(&oracle_data)) {
+                let oracle_data = option::destroy_some(oracle_data);
+                if (data::timestamp(&oracle_data) > min_timestamp) {
+                    vector::push_back(&mut values, *data::value(&oracle_data));
+                    vector::push_back(&mut oracles, *data::oracle_address(&oracle_data));
+                };
+            };
+        };
+        assert!(vector::length(&values) >= threshole, EValidDataSizeLessThanThreshold);
+        (values, oracles)
+    }
+
+    /// take the median value
+    public fun median<T: copy + drop>(meta_oracle: MetaOracle<T>): TrustedData<T> {
+        let (values, oracles) = combind(meta_oracle);
+        let sortedData = quick_sort(values);
```

Comment:
Probably worth having a generic quicksort function for others to use? Not sure how we can do that in move.

Reply:
I try to make this function generic by using bcs (not sure if that's the most efficient way, cc @damirka for comment), it now support all numeric types
and `DecimalValue`. I don't think there are easy way to make it support arbitrary types.
I will make it public for now so other can use it, my plan is to put this in a maths lib so more people can use it.